Source/WTF/wtf/FixedBitVector.h

```diff
@@ -0,0 +1,300 @@
+ * OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+ * (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+ * OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ */
+
+#pragma once
+
+#include <wtf/Assertions.h>
+#include <wtf/Atomics.h>
+#include <wtf/HashTraits.h>
+#include <wtf/PrintStream.h>
+#include <wtf/StdIntExtras.h>
+#include <wtf/StdLibExtras.h>
+
+namespace WTF {
+
+DECLARE_ALLOCATOR_WITH_HEAP_IDENTIFIER(FixedBitVector);
+DEFINE_ALLOCATOR_WITH_HEAP_IDENTIFIER(BitVector);
+
+class FixedBitVector final {
+    WTF_MAKE_FAST_ALLOCATED;
+    using WordType = UCPURegister;
+
+public:
+    FixedBitVector() = default;
+
+    ~FixedBitVector()
+    {
+        if (m_bits)
+            FixedBitVectorMalloc::free(m_bits);
+    }
+
+    bool concurrentAllocateOnce(size_t);
+    bool concurrentTest(size_t bitIndex, Dependency = Dependency());
+    bool concurrentTestAndSet(size_t bitIndex, Dependency = Dependency());
+
+    void allocateOnce(size_t);
+    bool testAndSet(size_t bitIndex);
+    bool test(size_t bitIndex);
+
+    size_t size() const { return m_size; }
+
+    size_t findBit(size_t startIndex, bool value) const;
+
+    bool operator==(const FixedBitVector&) const;
+
+    unsigned hash() const;
+
+    void dump(PrintStream& out) const;
+
+    class iterator {
+        WTF_MAKE_FAST_ALLOCATED;
+
+    public:
+        iterator()
+            : m_bitVector(nullptr)
+            , m_index(0)
+        {
+        }
+
+        iterator(const FixedBitVector& bitVector, size_t index)
+            : m_bitVector(&bitVector)
+            , m_index(index)
+        {
+        }
+
+        size_t operator*() const { return m_index; }
+
+        iterator& operator++()
+        {
+            m_index = m_bitVector->findBit(m_index + 1, true);
+            return *this;
+        }
+
+        iterator operator++(int)
+        {
+            iterator result = *this;
+            ++(*this);
+            return result;
+        }
+
+        bool isAtEnd() const
+        {
+            return m_index >= m_bitVector->size();
+        }
+
+        bool operator==(const iterator& other) const
+        {
+            return m_index == other.m_index;
+        }
+
+        bool operator!=(const iterator& other) const
+        {
+            return !(*this == other);
+        }
+
+    private:
+        const FixedBitVector* m_bitVector;
+        size_t m_index;
+    };
+
+    // Use this to iterate over set bits.
+    iterator begin() const
+    {
+        ASSERT(m_bits);
+        return iterator(*this, findBit(0, true));
+    }
+
+    iterator end() const
+    {
+        ASSERT(m_bits);
+        return iterator(*this, size());
+    }
+
+private:
+    bool get(size_t bitIndex, Dependency = Dependency()) const;
+
+    static constexpr unsigned wordSize = sizeof(WordType) * 8;
+    static constexpr WordType one = 1;
+
+    size_t m_size { 0 };
+    unsigned m_words { 0 };
+    WordType* m_bits { nullptr };
```

Comment:
I think we should use FixedVector for this storage. Then we can use m_bits.size() instead of m_words here. And it reduces the size of this class. Plus, it simplifies the implementation a lot. We do not need custom destructor (since FixedVector destructor destroys it). We do not need to have copy constructors since FixedVector will be copied. Probably we can make operator== simpler too.